### ISSUE TITLE
useAnimatedStyle gets out of sync with SharedValue modified inside useEffect on Android

### ISSUE THREAD

**Issue by gaearon** [software-mansion/react-native-reanimated#6677]
### Description

Here's the code.

```js
import React, {useState, useEffect} from 'react';
import {Pressable, Text} from 'react-native';
import Animated, {
  useAnimatedStyle,
  useSharedValue,
} from 'react-native-reanimated';

export default function App() {
  const [show, setShow] = useState(false);
  return show ? (
    <Lightbox onClose={() => setShow(false)} />
  ) : (
    <Pressable
      style={{padding: 100, fontSize: 50, backgroundColor: 'purple'}}
      onPress={() => setShow(true)}>
      <Text>open</Text>
    </Pressable>
  );
}

function Lightbox({onClose}) {
  const openProgress = useSharedValue(0);

  useEffect(() => {
     console.log('set to 1');
     openProgress.value = 1;
  }, [openProgress]);

  const containerStyle = useAnimatedStyle(() => {
    console.log(openProgress.value);
    if (openProgress.value < 1) {
      return {backgroundColor: 'red'};
    }
    return {backgroundColor: 'blue'};
  });

  return (
    <Animated.View
      style={[
        {
          position: 'absolute',
          top: 0,
          left: 0,
          right: 0,
          bottom: 0,
        },
        containerStyle,
      ]}>
      <Pressable style={{flex: 1}} onPress={onClose} />
    </Animated.View>
  );
}
```

### Steps to reproduce

1. Use Paper (this doesn't repro on Fabric)
2. Put the code above in your project
3. Run on Android

Repro case:

1. Press "open"
2. Tap on the opened blue screen (which will close it)
3. Press "open" again

Expected: the value will change `0 -> 1` and the screen will again be blue

Actual: the value will change `0 -> 1` **but the screen will remain red**

This only reproduces on Android Paper.

Wrapping `.value` assignment in `setTimeout` with `1` seems to "fix" it.

https://github.com/user-attachments/assets/9d3a42f2-a4ae-459d-a5e5-507034a74837

### Snack or a link to a repository

https://github.com/gaearon/reanimated-repro-weird

### Reanimated version

3.16.1

### React Native version

0.76.1, but repros on 0.74.1 so is not new

### Platforms

Android

### JavaScript runtime

Hermes

### Workflow

React Native

### Architecture

Paper (Old Architecture)

### Build type

Debug app & dev bundle

### Device

Android emulator

### Acknowledgements

Yes

**Comment by gaearon** [comment]
This doesn't seem to repro with ^3.15.5

**Comment by gaearon** [comment]
Okay the issue exists _at least_ as far back as `3.11.0`, so maybe this never even worked.

**Comment by joe-sam** [comment]
_SharedValues_ don't trigger a re-render and _useEffect_ runs after a render.  

You might consider doing one of two things.
Update the style (non-animated) in the _onPress_ handler before updating the _show_ state to trigger a re-render.

Run an animation based on the value _openProgress_ to interpolate styles, and not inside the _useEffect_ which is too late in the life cycle. Use the finishing callback of the animation (e.g. _withTiming_) if you don't want render styles on an interpolated value.

**Comment by joe-sam** [comment]
Does the useAnimatedStyle still not refresh the style with the new updated values, when a render takes place due to change in state or props ?

**Comment by gaearon** [comment]
You can see that the behavior is a bug by the discrepancy between iOS and Android, at the very least it is undefined behavior. But also, it's pretty established that it's enough to set `.value` to update `useAnimatedStyle`.

>When you change the sv.value Reanimated will update the styles and keep the shared value in sync between the threads. However, this won't trigger a typical React re-render because a shared value is a plain JavaScript object.

This is exactly the guarantee being broken in my repro case (only on Android Paper).

**Comment by amadeus** [comment]
I think you may have found the source of a bug that's been a real pain for us to figure out. In the Discord mobile app, in the voice ui, we had a situation where the control bar buttons would animate in with an opacity change, and in certain very hard to reproduce cases, they would never show up (but they would be physically mounted). Many of these shared values were being updated in useEffects as well, to trigger these mount animations.

Thank you finding a consistent repro!

**Comment by joe-sam** [comment]
I think the issue simply was that you were producing a "new " instance of the function component on every render and ended up recreating and discarding the shared value. 
 
 Anyway this reproducer has shed a lot of light on my own issues with the usage of  _useAnimatedStyle_, so thanks for the explanations.
 
 I have altered the code a bit so the background changes from red to blue after approx 10 taps. 
There is some weird accumulating floating point error. But that probably needs another discussion !!!

See [snack](https://snack.expo.dev/@joe-sam/useanimatedstyle-demo-fix)

**Comment by gaearon** [comment]
>I think the issue simply was that you were producing a "new " instance of the function component on every render and ended up recreating and discarding the shared value.

I _do_ create a new instance when the boolean state in the top-level component changes. However, that does not happen during the lifecycle of the inner component (before and after the effect), and it's the inner component where the problem happens.

With respect, I used to work on React, and I have a decent idea of how it works. The issue here is a Reanimated bug.

**Comment by tjzel** [comment]
I made a workaround for this problem. I'm not certain if this should be considered a bug on our side. I'll consult this with our team and folks from Meta to get more insight on it.

**Comment by tjzel** [comment]
Closing the issue since the Old Architecture is no longer supported.

**Comment by Maroka745** [comment]
TouchToRecord - Like Vine Instagram Whatsapp Video Recorder
This code just provide a way to record audio and video separately and then merge them both using ffmpeg command, so that we can have pause/record function for user.

Touch/Press android camera surface view to record video using FFmpeg Recorder from JavaCV. Using this I have implemented Vine like touch and hold video recording feature.

It will stop recording the video when user will lift the finger from camera surface view. User can pause and continue video recording after any interval of time. Once user have finished recording the video, click next to play the recorded video.

JavaCv Credits: https://github.com/bytedeco/javacv

LICENSE:

Give Credits and add this repo url to your blog/repo. No License restriction from me but JAVACV license has to be obeyed as I have made this code with help of that repo. Check that at the time you are using this code.

**Comment by Maroka745** [comment]
Notice: MediumMenu is no longer being maintained/updated.

MediumMenu
Medium 1.8.168 menu in Swift. That is still one of my favorite menus because that is easy to use and looks beautiful.

Inspired by RBMenu. I made that a reference and customized a fine point.

Demo

MediumMenu

**Comment by Maroka745** [comment]
React Native
Learn once, write anywhere:
Build mobile apps with React.

 React Native is released under the MIT license.  Current CircleCI build status.  Current npm package version. PRs welcome!  Follow @reactnative

Getting Started · Learn the Basics · Showcase · Contribute · Community · Support

React Native brings React's declarative UI framework to iOS and Android. With React Native, you use native UI controls and have full access to the native platform.

Declarative. React makes it painless to create interactive UIs. Declarative views make your code more predictable and easier to debug.
Component-Based. Build encapsulated components that manage their state, then compose them to make complex UIs.
Developer Velocity. See local changes in seconds. Changes to JavaScript code can be live reloaded without rebuilding the native app.
Portability. Reuse code across iOS, Android, and other platforms.
React Native is developed and supported by many companies and individual core contributors. Find out more in our ecosystem overview.

Contents

Requirements
Building your first React Native app
Documentation
Upgrading
How to Contribute
Code of Conduct
License
📋 Requirements

React Native apps may target iOS 10.0 and Android 4.1 (API 16) or newer. You may use Windows, macOS, or Linux as your development operating system, though building and running iOS apps is limited to macOS. Tools like Expo can be used to work around this.

🎉 Building your first React Native app

Follow the Getting Started guide. The recommended way to install React Native depends on your project. Here you can find short guides for the most common scenarios:

Trying out React Native
Creating a New Application
Adding React Native to an Existing Application
📖 Documentation

The full documentation for React Native can be found on our website.

The React Native documentation discusses components, APIs, and topics that are specific to React Native. For further documentation on the React API that is shared between React Native and React DOM, refer to the React documentation.

The source for the React Native documentation and website is hosted on a separate repo, @facebook/react-native-website.

🚀 Upgrading

Upgrading to new versions of React Native may give you access to more APIs, views, developer tools, and other goodies. See the Upgrading Guide for instructions.

React Native releases are discussed in the React Native Community, @react-native-community/react-native-releases.

👏 How to Contribute

The main purpose of this repository is to continue evolving React Native core. We want to make contributing to this project as easy and transparent as possible, and we are grateful to the community for contributing bug fixes and improvements. Read below to learn how you can take part in improving React Native.

Code of Conduct

Facebook has adopted a Code of Conduct that we expect project participants to adhere to. Please read the full text so that you can understand what actions will and will not be tolerated.

Contributing Guide

Read our Contributing Guide to learn about our development process, how to propose bugfixes and improvements, and how to build and test your changes to React Native.

Open Source Roadmap

You can learn more about our vision for React Native in the Roadmap.

Good First Issues

We have a list of good first issues that contain bugs which have a relatively limited scope. This is a great place to get started, gain experience, and get familiar with our contribution process.

Discussions

Larger discussions and proposals are discussed in @react-native-community/discussions-and-proposals.

📄 License

React Native is MIT licensed, as found in the LICENSE file.

React Native documentation is Creative Commons licensed, as found in the LICENSE-docs file.

**Comment by Maroka745** [comment]
Merge pull request GeekyAnts#26 from stdavis/patch-1

**Comment by Maroka745** [comment]
IoT Car Share System
An automatic car share system which is used to book, find, unlock and lock a car. There are four types of users which use this application:

Customers: can register, log-in, search and book cars. A booking will automatically be added to the customer's Google Calendar. They can also unlock a booked-car using a console-based system or facial recognition AI.

System administrators: can view car rental history, users, cars, raise an issue with a car for engineers to investigate.

Company managers: can view traffic related metrics to make better business decision, such as Daily active users, percentage of usage for each car per day.

Engineers: are responsible for repairing cars. Bluetooth is used to automatically unlock a car when an engineer is near it, and the engineer's QR code is used to store history of maintenance

Pre-requisites

Raspberry PI model 4
Python v3.6+
Flask (Python framework)
Google Calendar API access
GCP IoT access
Architecture

The high level architecture diagram of the app is as follows: Architecture Diagram

**Comment by Maroka745** [comment]
Job match making system
📝 PROJECT IDEA

Build an end-to-end job matching portal helping hiring-managers to find their ideal candidate and job seekers to find their dream job and allowing hiring managers to conduct one on one interview within the app through video conferencing feature.

📝 FEATURES

(1) Allow users to sign up and login to the application

Employer / Hiring Manager Functionality (2) Allow employers to view their job listings and how each job is performing
(3) Allow employers to view the number of matched candidates vs total number of candidates using a line chart

(4) Allow employers to create , edit and delete a job posting

(5) Allow employers to view applicants of each job listing

(6) Allow employers to download applicants resume

(7) Allow employers to join or create a video interview session within the same app

(8) Provide employer with matching percentage of a candidate (Using ML Techniques)

Sample Employer User credentials:

Email: prabhav@mehra.com

Password: Password12@

Employee/ Job Seeker Functionality (1) Allow employees to view all the job listings

(2) Allow employee to filter job postings based on location or job type

(3) Allow employee to upload their resume so they can apply for jobs

(4) Allow employee to join a video interview session

Sample Employee User credentials:

Email: ali@aldi.com

Password: Password12@

Admin Dashboard (Required by Client) (1) Allow admin to view, edit or delete all job postings

(2) Allow admin to view all job applicants

Sample Admin User credentials:
Email: admin@admin.com
Password: Password12@
Frontend (ReactJS)
Frontend folder contains UI for all pages such as Sign In, Sign Up, Employee dashboard, Employer ashboard and many more. The pages are connected through the backend (Python) using restful APIS, which were done through AXIOS, are loaded upon request through the axios requests.
NPM version used: 7.11.2

To run the frontend, you have to run the following commands:

$ npm install
$ npm start
Backend (Python)
Backend contains restful APIS which triggers lambda functions. Backend uses AWS dynamodb and s3 as a database. It also authenticates a user using AWS Cognito.
Python Version used: 3.7.7 Flask Version used: 1.1.2

To run the backend, you have to run the following commands:

$ pip3 install -r requirements.txt
$ python3 main.py
Github URL:
https://github.com/kaushal-gawri9899/Programming-Project-1.git

Application URL:
http://programming-frontend-deployment.s3-website-us-east-1.amazonaws.com/

**Comment by Maroka745** [comment]
Ownego JS Dev Test
🚀 Project Overview

Đây là một ứng dụng web nhỏ giúp quản lý các quán trà sữa tại Nam Đồng. Ứng dụng hiển thị danh sách quán, menu các loại trà sữa và hỗ trợ lọc, sắp xếp theo các tiêu chí khác nhau.

✨ Features

✅ Hiển thị danh sách quán trà sữa từ dữ liệu JSON.
✅ Hiển thị menu đồ uống của từng cửa hàng.
✅ Hỗ trợ sắp xếp (Sort) theo:

Tên (A → Z, Z → A)
Giá (Thấp → Cao, Cao → Thấp)
✅ Hỗ trợ lọc (Filter) theo topping như yêu cầu UI 1b.
✅ Responsive theo các breakpoints: 768px, 360px.
✅ Sidebar có thể mở/đóng trên mobile.
✅ Mock API Backend bằng Node.js + Express.

🛠️ Installation & Setup

1️⃣ Clone Project

git clone https://github.com/trungvu2011/ownego-js-test.git
cd ownego-dev-test
2️⃣ Install Dependencies

npm install
3️⃣ Start Mock API Server

cd backend
node server.js
Server chạy trên: http://localhost:8080
4️⃣ Start React App

npm run dev
Ứng dụng chạy trên: http://localhost:5173
📝 Author

Vũ Đức Trung

Email: trungvd@gmail.com
GitHub: trungvd
Made with ❤️ by Vũ Đức Trung 🚀

**Comment by Maroka745** [comment]
Hi 👋, I'm Taiwo Damilola
tdamm92

📫 How to reach me tdammy92@gmail.com
🧰 Take a pick at my portfolio
Connect with me:

 akshay113am  akshay k

Languages and Tools:

 reactnative  react  NextJS  javascript  typescript  html5 css3  tailwind  electron  electron

 firebase  firebase  firebase  mongodb  mongodb  mongodb swift  android  docker  docker

**Comment by Maroka745** [comment]
Build a Kanban Board with Javascript
Status

Published

Overview

This project is design to be completed on Pluralsight. To find out more see here: https://www.pluralsight.com/product/projects.

Installation

Run the following command from the root folder of the cloned project to install all dependencies.

npm install

Verify Setup

In order to verify that everything is setup correctly, run the following command, which should show you the failing tests. This is good! We'll be fixing these tests once we jump into the build step.

npm run test

Every time you want to check your work locally you can type that command, and it will report the status of every task in that module.

As you move through the modules, you can run module-specific tests with the script npm run test:module1, replacing the number with one that corresponds with the module you are working in.

Previewing Your Work

In order to see your changes in a browser, you can run npm start from the command line. This will open a browser and you should see your landing page.

**Comment by Maroka745** [comment]
#Buy-Now Front end app

This project was bootstrapped with Create React App.

**Comment by Maroka745** [comment]
Vaccine SlotChecker
App logo

Check available vaccine sessions/slots at your nearest vaccination center

NOTE : This app is only made for educational purpose as a part of my learning. In case of removal or any issue please contact me.

This app will provide most recent vaccination slot availability through government provided facility Co-Win.
The user needs to feed in the pincode or district name to receive information related to vaccination centres around them .
This app doesn't require login. If slots are available then the user can book their slot by clicking the sign in/register button
which, will take them to the official Co-Win website where they can register themselves and their family members.

This app provides data using the public API provided by cowin and and i am not directly or indirectly responsible for the data provided in case of any mismatch

As per GOVT policy of CoWIN the appointment availability data is chached and may be up to 5 minutes old .

For more information related to this API please refer to Co-WIN Public APIs

Icons made by ultimatearm and Freepik from www.flaticon.com

Screenshots

ss1	ss1	
ss1	ss1	
Permissions

INTERNET : This app requires internet connectivity to fetch most recent data and serve it to the users.
Supported devices

Any mobile phone runing on Android operating system is able to run this application.

NOTE : This app is not available for ios devices.

Contributing Guidelines

First fork this repository

Then in your terminal type :

$ git clone https://github.com/username/Vaccine-SlotChecker-Android

NOTE : Above here replace username with your github username

Now you will have a local copy of your version of the original repository.

Now change the directory

$ cd Vaccine-SlotChecker-Android

Add a conection to the original repository

$ git remote add upstream https://github.com/abhishek-55/Vaccine-SlotChecker-Android

Confirm by checking the upstream

$ git remote -v

Make your own banch with your username

$ git checkout "branch-name"

Now make changes in the files

git add and git commit those changes and git push them back to github.

These will go to your version of the repository.

Now create a Pull Request (PR)

Go to your version of the repository on github.Click the “New pull Request” button at the top.
Note that original repository will be on the left and your forked repository will be on the right.

Click the green button “Create pull request”. Give a succinct and informative title, in the comment field give a short explanation of the changes and click the green button “Create pull request” again.

Pulling others' changes

Before you make further changes to the repository, you should check that your version is even with the upstream or not.

$ git pull upstream development --rebase

LICENSE

The application is licensed under the MIT license.

**Comment by Maroka745** [comment]
IP Activities Assignment
This is a take-home assignment for evaluating senior Ruby on Rails developers. The assignment focuses on IP activities functionality.

Background

The application tracks IP activities for users and their trading accounts. Each IP activity has an activity type (trade, login, kyc) and is associated with either a user or a trading account.

The database contains 100M records, so performance optimization is crucial.

Assignment

Task 1: Optimize IpActivity Scopes and Implement Smart Limiting

Context:

User has related TradingAccount model
Both have related IpActivity model that logs user activity
Each IpActivity has its own type:
kyc - IP addresses from which user passes verification, no more than 10 records
login - IP addresses from which user logs in, these activities are tied to user, no more than 1000 records
trade - activities that log IP addresses of client trading operations from different platforms, can be 1000-10000 records
Task: Display in the index controller action for further serialization and map rendering the latest user activity by date. Regardless of date, ensure that kyc and login types are always displayed (no more than 1000 per each type) and the remaining slots up to a total limit of 3000 are filled with trade activities.

Example: If a client has IP activities with types: kyc - 1, login - 500, trade - 2000 Should output: 1 + 500 + 1000 = 1501 IpActivity records

Requirements:

Optimize the existing scopes in the IpActivity model
Implement smart limiting logic that prioritizes kyc and login activities
Fill remaining slots with trade activities up to total limit of 3000
Consider database performance with 100M records
Maintain existing API functionality
Current scopes to optimize:

  scope :for_user, lambda { |user|
    where(user:).includes(:user, :trading_account, :ip_address_record)
                .or(where(trading_account_login: user.trading_accounts.select(:login)))
  }
  scope :recent_n_per_activity_type, lambda { |limit|
    query = <<-SQL.squish
      SELECT activity_limited.id
      FROM (SELECT DISTINCT activity_type FROM ip_activities) activity_groups
      JOIN LATERAL (
        SELECT * FROM ip_activities activity_all
        WHERE activity_all.activity_type = activity_groups.activity_type
        ORDER BY activity_all.created_at DESC
        LIMIT :limit
      ) activity_limited ON true
    SQL

    where("ip_activities.id IN (#{ApplicationRecord.sanitize_sql([query, { limit: }])})")
  }
Task 2: Design Reusable Filter Component Architecture

Context: The current implementation lacks a filtering component. The task is to design the architecture of a reusable filter component that can accept parameters, save them to the database, and be universal enough to be applied to filtering users, trading_accounts, or other models with their own parameter sets.

Requirements:

Design a universal filter model that can be applied to different entities (users, trading_accounts, ip_activities)
Filter should work in both index and filter_metadata endpoints
Parameters should be persistable in the database
Component should be reusable across different controllers and models
Support different filter types (date ranges, enums, text search, etc.)
Deliverables: Choose one of the implementation options:

Option A: Complete code implementation with all necessary models, controllers, and services
Option B: Detailed architecture description with schema design, usage examples, and implementation sketches
Consider:

How to handle different parameter types for different models
Database schema design for storing filter configurations
API design for filter endpoints
Performance implications
Code reusability and maintainability
Getting Started

Clone the repository
Run bundle install
Set up the database: rails db:create db:migrate db:seed
Start the server: rails s
API Endpoints

GET /api/users/:user_id/ip_activities - Returns IP activities for a user
GET /api/users/:user_id/ip_activities/filter_metadata - Returns metadata for filtering IP activities
Testing

Run the tests with:

bin/rspec
Submission

Please provide your optimized solution along with an explanation of your approach and any trade-offs you made.

**Comment by Maroka745** [comment]
Frontend Trading Terminal Challenge
Overview

This is a senior/lead frontend developer challenge designed to test core skills required for building real-time trading applications. The challenge focuses on WebSocket integration, real-time data handling, and React/TypeScript proficiency.

Important: Approach this task as if it will be used in production by real users. A certain degree of ownership and attention to detail is expected.

Challenge Requirements

Build a Real-Time Cryptocurrency Pairs List that displays live market data with detailed symbol information.

Core Features (Required)

1. Trading Pairs List

Fetch and display available trading pairs from Binance REST API
Show initial list of at least 20-30 major USDT pairs (BTC/USDT, ETH/USDT, etc.)
Display in a clean table/list format with columns:
Symbol (e.g., BTCUSDT)
Current Price
24h Change % (with color coding: green for positive, red for negative)
24h Volume
2. Real-Time Price Updates

Connect to Binance WebSocket API for live price tickers
Update prices in real-time without page refresh
Implement proper WebSocket connection management:
Connection status indicator
Automatic reconnection on disconnect
Error handling for connection failures
3. Symbol Detail Popup/Modal

Click on any trading pair to open detailed information
Display in a modal/popup/sidebar:
Symbol details (base/quote currencies)
Current price (real-time)
24h statistics: high, low, volume, change %
Price precision and minimum order quantity
Real-time price updates within the modal
4. Basic UI/UX Features

Search/filter functionality for trading pairs
Loading states while fetching initial data
Error handling with user-friendly messages
Technical Requirements

Core Tech Stack

React 18 with TypeScript
Vite for build tooling
State management of your choice (React hooks, Redux Toolkit, Zustand, etc.)
Styling solution of your choice (CSS modules, Tailwind, styled-components, etc.)
WebSocket Integration

Native WebSocket or any WebSocket library
Handle multiple symbol subscriptions efficiently
Proper cleanup on component unmount
Connection resilience (reconnect logic)
Code Quality

TypeScript interfaces for all API responses
Error boundaries for graceful error handling
Custom hooks for WebSocket and data management
Clean component structure with separation of concerns
API Resources

REST API - Get Trading Pairs

GET https://api.binance.com/api/v3/exchangeInfo
Returns all available trading pairs and their details
Filter for USDT pairs for simplicity
No API key required
WebSocket API - Real-time Prices

wss://stream.binance.com:9443/ws/!ticker@arr
Streams all 24hr ticker statistics for all symbols
Real-time updates every second
No API key required
Alternative single symbol ticker:

wss://stream.binance.com:9443/ws/btcusdt@ticker
Getting Started

Prerequisites

Node.js 18+ and npm/yarn
Basic understanding of WebSocket APIs
Quick Setup

Install dependencies:

npm install
Start development:

npm run dev
Test API endpoints:

# Test REST API
curl https://api.binance.com/api/v3/exchangeInfo

# Test WebSocket (in browser console)
const ws = new WebSocket('wss://stream.binance.com:9443/ws/btcusdt@ticker');
ws.onmessage = (event) => console.log(JSON.parse(event.data));
Evaluation Criteria

Technical Implementation (40%)

WebSocket connection management and error handling
Real-time data updates efficiency
TypeScript usage and type safety
Component architecture and reusability
Code Quality (35%)

Clean, readable, and maintainable code
Proper error handling and edge cases
Performance considerations
Code organization and structure
User Experience (25%)

Intuitive and functional interface
Smooth real-time updates
Loading states and error messages
Modal/popup implementation
Note on Design: We will not judge visual design or aesthetics, but the application should be reasonably well-presented. Any significant UX/UI flaws that impact usability will be considered in the evaluation.

Time Expectations

Minimum viable solution: 3-4 hours
Complete implementation: 6-8 hours
Polished with extras: 10+ hours
Deliverables

Complete source code with clear Git commits
README with setup instructions and architecture notes
Working application that runs with npm run dev and npm run build
TypeScript properly configured and used throughout
Bonus Points (Optional)

Performance optimization for handling many real-time updates
Advanced filtering/search with debouncing
Price change animations (flash green/red on updates)
Virtualized list for handling large datasets
Unit tests for key components
Error monitoring or logging implementation
Focus on building a production-quality solution that demonstrates your ability to handle real-time financial data effectively.

Good luck! 🚀

**Comment by Maroka745** [comment]
Please add more cliparts to the clipart screen and enable to include several cliparts in the badge view itself.

Screenshots for the change:
iss_240

**Comment by Maroka745** [comment]
Fludget
code

Browse through a variety of widgets used in flutter and learn how to use them

 made-by-acm  license Awesome Flutter

Project Status

stars  forks issuses prs

This application is developed to learn Flutter using Flutter. Different widgets used in Flutter can be viewed in this app along with their implementation, description and code. If you like our work, ⭐ the repository to show support.

Screenshots of the application
Splash Screen



Home Screen

Light Mode

  
Dark Mode

  
Implementation Example

**Comment by Maroka745** [comment]
Kicking Off Hacktoberfest with ACM-VIT!


Golly Bot!

A general purpose Discord bot, written in GO!

 made-by-acm

Submitting a Pull Request

Fork the repository by clicking the fork button on top right corner of the page
Clone the target repository. To clone, click on the clone button and copy the https address. Then run
git clone https://github.com/ACM-VIT/golly-bot.git
Go to the cloned directory by running
cd golly-bot
Create a new branch. Use
 git checkout -b mynewbranch
Make your changes to the code. Add changes to your branch by using
git add .
Commit the chanes by executing
git commit -m "short message describing changes"
Push to remote. To do this, run
git push origin mynewbranch
Create a pull request. Go to the target repository and click on the "Compare & pull request" button. Make sure your PR description mentions which issues you're solving. Wait for your request to be accepted, and you're good to go!


Guidelines for Pull Request

Avoid pull requests that :
are automated or scripted
that are plagarized from someone else's branch
Do not spam
Project maintainer's decision on validity of PR is final.
For additional guidelines, refer to this website for additional information

Instructions on how to run the bot locally

In the cloned directory :
Add a .env file following the .env.sample file provided and add the required secrets like token,etc.
(Instructions on how to create a Discord Bot token can be found here)
Allow the bot to have Privilaged Intents as found here
Make sure you have go and all its dependencies installed in your system.
(Instructions can be found here)
Run the following command
go run main.go
Instructions on how to run the bot in a docker container

Build the docker image from the Dockerfile using the following command, make sure you are inside the project folder.
docker build -t <your image name> .
Now run the application inside the container using the following command
docker run -it <your image name>
To specify command line arguments, use
docker run -it <your image name> <arguments>
For example:
docker run -it <your image name> -rmcmd=false
License

License

Authors

Anish Raghavendra
Manav Muthanna
Sarthak Gupta

**Comment by Maroka745** [comment]
Covid-19-Tracker-for-India
An Innovative app that helps you analyse the Covid situation in the country. It shows case statistics through figures and also graphically. The innovative feature is that it also shows extent of impact in various states via map overlay. Also the latest Covid related News and Guidelines are available.

Screenshots

**Comment by Maroka745** [comment]
Quizzie App Companion (Android)
Android client for Quizzie

    Android-master Actions Status    License: MIT     Figma Designs

Quizzie
This project is a complete online quiz management platform with a plethora of features for teachers and students. It provides all the basic functionalities of a quiz portal like timer, scoring, ranking etc. alongside analysis of results from the students on the teacher dashboard, summary of tests for the student, Login and signup using a Google account for both teacher and student. It provides a smooth user interface for both the teachers and students to create and attempt different kinds of quizzes seamlessly.

Quizzie Android app allows as an extension to the organizer's side of the website to allow them to add questions by capturing it from a book or elsewhere just to make the process of adding questions simpler for teachers

DOCS

Motivation

The motivation of this project came from the idea to help teachers and institute administrators create quizzes and tests in an efficient manner. With the popularity of the Internet, it is inevitable to have online quizzes as classroom assessments. In learning, online quizzes may serve two objectives, that is, for self-study or as a formal assessment. The online quiz has its advantages, such as saving the cost of paper printing and reducing the time spent for having assessments in class. However, there are weaknesses. Its primary drawback is the issue of academic dishonesty, especially when students are answering online quiz.

Android APK - Open this link in the Android device to install the app

Screenshots

Quizzie Screenshots

How is the app built?

This app is coded natively using MVVM(Model View - View Model) architecture in Kotlin. It is built in Android Studio.

Structure

build.gradle - root gradle config file
settings.gradle - root gradle settings file
app - our only project in this repo
app/build.gradle - project gradle config file
app/src - main project source directory
app/src/main - main project flavour
app/src/main/AndroidManifest.xml - manifest file
app/src/main/java - java source directory
app/src/main/res - resources directory
Features

An organizer can create a quiz, add questions to it and publish them
Organizer can add questions from a book by capturing it from the Android device
The text recognition module automatically captures the text from the image
There is no capture button to capture the image. It automatically captures after the text is recognized
The question can be edited anytime and the changes to the text can be entered manually as well.
The quiz after it is saved is made available for the students to attend from the website.
Execution

This github repositority can be cloned using "git clone <github_link>" command
Android Studio has to be installed (Version > 4.0)
The cloned repo in the local PC should be opened using Android Studio
Follow the unpacking instructions below to execute the build and run the app
Unpacking

This repository uses secrets that aren't pushed to the version control. However they are needed to build the project. Follow the below steps to obtain the files.

From the project root, execute:

.github/encrypted_secrets/decrypt_secrets.sh SECRET_KEY
If you are not on the development team contact us at amit.ananthkumar@gmail.com to obtain the SECRET_KEY used to unlock the secrets.

Building

It is recommended that you run Gradle with the --daemon option, as starting up the tool from scratch often takes at least a few seconds. You can kill the java process that it leaves running once you are done running your commands.

Tasks work much like Make targets, so you may concatenate them. Tasks are not re-done if multiple targets in a single command require them. For example, running assemble install will not compile the apk twice even though install depends on assemble.

Clean

gradle clean
Debug

This compiles a debugging apk in build/outputs/apk/ signed with a debug key, ready to be installed for testing purposes.

gradle assembleDebug
You can also install it on your attached device:

gradle installDebug
Release

This compiles an unsigned release (non-debugging) apk in build/outputs/apk/. It's not signed, you must sign it before it can be installed by any users.

gradle assembleRelease
Test

Were you to add automated java tests, you could configure them in your build.gradle file and run them within gradle as well.

gradle test
Lint

This analyses the code and produces reports containing warnings about your application in build/outputs/lint/.

gradle lint
Contributors

Amit Krishna A
Amit Krishna A (Insert Your Image Link In Src

  

Dinesh B S
Dinesh B S (Insert Your Image Link In Src

  

Further reading

Build System Overview
Gradle Plugin User Guide
Gradle Plugin Release Notes

**Comment by Maroka745** [comment]
Mend Backend
Go Build & Test

AI-powered backend for Mend — Couples Therapy Platform 🚀

**Comment by Maroka745** [comment]
ALLXFR
AXFR all the things!



This tool performs a zone transfer (AXFR)against the root zone servers to obtain the root zone file. And then attempts opportunistic zone transfers for every IP for every nameserver for a given zone. Additionally, each NS/A/AAAA record is also re-queried to find additional servers or IPs not included as root glue.

Most zones do not allow zone transfers, however a few do. Sometimes only on a single IP for a given nameserver and not the others, and sometimes only for servers or IPs that are authoritative but not included in the root zones. This tool will try them all and save every successful transfer.

This tool works best on an IPv4/IPv6 dual stack internet connection.

Providing a zone file with the -zonefileflag will attempt a transfer with the domains and sub-domains in the zone file provided.

TLDs in the Public Suffix List can be attempted as well with the -psl flag.

Example

./allxfr -dry-run
ROOT g.root-servers.net. xfr size: 22017 records in 1.334s
mr. ns-mr.nic.tn. (41.228.63.70) xfr size: 444 records in 337ms
sl. ns1.neoip.com. (45.83.41.38) xfr size: 455 records in 592ms
sy. ns1.tld.sy. (82.137.200.85) xfr size: 1594 records in 870ms
cd. ns-root-2.scpt-network.com. (41.79.235.67) xfr size: 497 records in 598ms
my. a.nic.my. (103.44.108.53) xfr size: 1592 records in 608ms
mq. ns1-mq.mediaserv.net. (213.16.20.3) xfr size: 1541 records in 299ms
td. nsa.nic.td. (154.68.159.246) xfr size: 492 records in 1.821s
gp. ns2.nic.gp. (193.218.114.34) xfr size: 1578 records in 195ms
xn--54b7fta0cc. bayanno.btcl.net.bd. (180.211.212.213) xfr size: 915 records in 728ms
xn--ogbpf8fl. ns1.tld.sy. (82.137.200.85) xfr size: 271 records in 460ms
xn--90ae. a.nic.bg. (2a02:6a80::192:92:129:99) xfr size: 305 records in 326ms
cv. cv01.dns.pt. (185.39.208.18) xfr size: 507 records in 333ms
bd. surma.btcl.net.bd. (203.112.194.232) xfr size: 52 records in 486ms
rw. ns1.ricta.org.rw. (196.49.7.188) xfr size: 1515 records in 1.269s
gf. ns1-mq.mediaserv.net. (213.16.20.3) xfr size: 1531 records in 304ms
bn. ns2.bnnic.bn. (202.93.214.163) xfr size: 79 records in 395ms
tj. ns2.tojikiston.com. (193.111.11.4) xfr size: 1690 records in 998ms
pg. ns1.tiare.net.pg. (202.165.192.23) xfr size: 447 records in 409ms
mw. domwe.sdn.mw. (41.87.5.162) xfr size: 1538 records in 938ms
xn--j1amh. tier1.num.net.ua. (193.110.163.134) xfr size: 1055 records in 1.349s
sv. cir.red.sv. (168.243.254.1) xfr size: 1514 records in 883ms
21 / 1516 transfered in 3m29.92
Usage

Usage of ./allxfr:
  -dry-run
        only test if xfr is allowed by retrieving one envelope
  -ixfr
        attempt an IXFR instead of AXFR
  -out string
        directory to save found zones in (default "zones")
  -overwrite
        if zone already exists on disk, overwrite it with newer data
  -parallel uint
        number of parallel zone transfers to perform (default 10)
  -psl
        attempt AXFR from zones listed in the public suffix list
  -retry int
        number of times to retry failed operations (default 3)
  -save-all
        attempt AXFR from every nameserver for a given zone and save all answers
  -status-listen string
        enable HTTP status server on specified [IP:]port (e.g., '8080', '127.0.0.1:8080', '[::1]:8080')
  -verbose
        enable verbose output
  -version
        print version and exit
  -zonefile string
        use the provided zonefile instead of getting the root zonefile
Building

go build
extsort
PkgGoDev Go Report Card

A high-performance external sortinglibrary for Go that enables sorting of arbitrarily large datasets that don't fit entirely in memory. The library operates on channels and uses temporary disk storage to handle datasets larger than available RAM.

Features

Memory Efficient: Sorts datasets of any size using configurable memory limits
High Performance: Optimized for throughput with parallel sorting and merging
Generic Support: Modern Go generics for type-safe operations
Legacy Compatible: Maintains backward compatibility with interface-based API
Cross-Platform: Works on Unix, Linux, macOS, and Windows
Channel-Based: Integrates seamlessly with Go's concurrency patterns
Installation

go get github.com/lanrat/extsort
Quick Start

IMPORTANT: All input channels must be closed after all data has been sent. The Sort() method will continue reading from the input channel until it is closed. Failure to close the input channel will cause the sorting process to hang indefinitely.

Generic API (Recommended)

The modern generic API provides type safety and improved performance:

import (
    "context"
    "fmt"
    "math/rand"

    "github.com/lanrat/extsort"
)

func main() {
    // Create input channel with unsorted integers
    inputChan := make(chan int, 100)
    go func() {
        defer close(inputChan)
        for i := 0; i < 1000000; i++ {
            inputChan <- rand.Int()
        }
    }()

    // Sort using the generic API
    sorter, outputChan, errChan := extsort.Ordered(inputChan, nil)

    // Start sorting in background
    go sorter.Sort(context.Background())

    // Process sorted results
    for value := range outputChan {
        fmt.Println(value)
    }

    // Check for errors
    if err := <-errChan; err != nil {
        panic(err)
    }
}
String Sorting

For string data, use the optimized string sorter:

import (
    "context"
    "fmt"

    "github.com/lanrat/extsort"
)

func main() {
    words := []string{"zebra", "apple", "banana", "cherry"}

    inputChan := make(chan string, len(words))
    for _, word := range words {
        inputChan <- word
    }
    close(inputChan)

    sorter, outputChan, errChan := extsort.Strings(inputChan, nil)
    go sorter.Sort(context.Background())

    fmt.Println("Sorted words:")
    for word := range outputChan {
        fmt.Println(word)
    }

    if err := <-errChan; err != nil {
        panic(err)
    }
}
Custom Types with Generic API

import (
    "bytes"
    "context"
    "encoding/gob"
    "fmt"

    "github.com/lanrat/extsort"
)

type Person struct {
    Name string
    Age  int
}

func personToBytes(p Person) ([]byte, error) {
    var buf bytes.Buffer
    enc := gob.NewEncoder(&buf)
    err := enc.Encode(p)
    return buf.Bytes(), err
}

func personFromBytes(data []byte) (Person, error) {
    var p Person
    buf := bytes.NewReader(data)
    dec := gob.NewDecoder(buf)
    err := dec.Decode(&p)
    return p, err
}

func comparePersonsByAge(a, b Person) int {
    // Sort by age
    if a.Age != b.Age {
        if a.Age < b.Age {
            return -1
        }
        return 1
    }
    return 0
}

func main() {
    people := []Person{
        {"Alice", 30},
        {"Bob", 25},
        {"Charlie", 35},
    }

    inputChan := make(chan Person, len(people))
    for _, person := range people {
        inputChan <- person
    }
    close(inputChan)

    sorter, outputChan, errChan := extsort.Generic(
        inputChan,
        personFromBytes,
        personToBytes,
        comparePersonsByAge,
        nil,
    )

    go sorter.Sort(context.Background())

    fmt.Println("People sorted by age:")
    for person := range outputChan {
        fmt.Printf("%s (age %d)\n", person.Name, person.Age)
    }

    if err := <-errChan; err != nil {
        panic(err)
    }
}
Configuration

Customize sorting behavior with the Config struct:

config := &extsort.Config{
    ChunkSize:          500000,  // Records per chunk (default: 1M)
    NumWorkers:         4,       // Parallel sorting/merging workers (default: 2)
    ChanBuffSize:       10,      // Channel buffer size (default: 1)
    SortedChanBuffSize: 1000,    // Output channel buffer (default: 1000)
    TempFilesDir:       "/var/tmp",  // Temporary files directory (default: intelligent selection)
}

sorter, outputChan, errChan := extsort.Ordered(inputChan, config)
Temporary Directory Selection

When TempFilesDir is empty (default), the library intelligently selects temporary directories that prefer disk-backed locations over potentially memory-backed filesystems. On Linux systems where /tmp may be mounted as tmpfs (memory-backed), this helps prevent out-of-memory issues when sorting datasets larger than available RAM.

For production use with large datasets, it's recommended to explicitly set TempFilesDir to a known disk-backed directory (such as /var/tmp on Unix systems) to ensure optimal performance and avoid memory limitations.

Legacy Interface-Based API

The library maintains backward compatibility with the original interface-based API:

import (
    "context"
    "encoding/binary"
    "fmt"
    "math/rand"

    "github.com/lanrat/extsort"
)

type sortInt struct {
    value int64
}

func (s sortInt) ToBytes() []byte {
    buf := make([]byte, 8)
    binary.LittleEndian.PutUint64(buf, uint64(s.value))
    return buf
}

func sortIntFromBytes(data []byte) extsort.SortType {
    value := int64(binary.LittleEndian.Uint64(data))
    return sortInt{value: value}
}

func compareSortInt(a, b extsort.SortType) bool {
    return a.(sortInt).value < b.(sortInt).value
}

func main() {
    inputChan := make(chan extsort.SortType, 100)
    go func() {
        defer close(inputChan)
        for i := 0; i < 100000; i++ {
            inputChan <- sortInt{value: rand.Int63()}
        }
    }()

    sorter, outputChan, errChan := extsort.New(
        inputChan,
        sortIntFromBytes,
        compareSortInt,
        nil,
    )

    go sorter.Sort(context.Background())

    for item := range outputChan {
        fmt.Println(item.(sortInt).value)
    }

    if err := <-errChan; err != nil {
        panic(err)
    }
}
Diff Sub-Package

The diff sub-package provides functionality for comparing two sorted data streams and identifying differences. It's particularly useful for comparing large datasets efficiently.

Basic String Diff

import (
    "context"
    "fmt"

    "github.com/lanrat/extsort/diff"
)

func main() {
    // Create two sorted string channels
    streamA := make(chan string, 5)
    streamB := make(chan string, 5)

    // Populate stream A
    go func() {
        defer close(streamA)
        for _, item := range []string{"apple", "banana", "cherry", "elderberry"} {
            streamA <- item
        }
    }()

    // Populate stream B
    go func() {
        defer close(streamB)
        for _, item := range []string{"banana", "cherry", "date", "fig"} {
            streamB <- item
        }
    }()

    // Create error channels
    errA := make(chan error, 1)
    errB := make(chan error, 1)
    close(errA)
    close(errB)

    // Process differences
    result, err := diff.Strings(
        context.Background(),
        streamA, streamB,
        errA, errB,
        func(delta diff.Delta, item string) error {
            switch delta {
            case diff.OLD:
                fmt.Printf("Only in A: %s\n", item)
            case diff.NEW:
                fmt.Printf("Only in B: %s\n", item)
            }
            return nil
        },
    )

    if err != nil {
        panic(err)
    }

    fmt.Printf("Summary: %d items only in A, %d items only in B, %d common items\n",
        result.ExtraA, result.ExtraB, result.Common)
}
Generic Diff for Custom Types

import (
    "context"
    "fmt"

    "github.com/lanrat/extsort/diff"
)

func main() {
    // Create channels with integer data
    streamA := make(chan int, 5)
    streamB := make(chan int, 5)
    errA := make(chan error, 1)
    errB := make(chan error, 1)

    // Populate streams
    go func() {
        defer close(streamA)
        defer close(errA)
        for _, num := range []int{1, 3, 5, 7, 9} {
            streamA <- num
        }
    }()

    go func() {
        defer close(streamB)
        defer close(errB)
        for _, num := range []int{2, 4, 5, 6, 8} {
            streamB <- num
        }
    }()

    // Compare using generic diff
    compareFunc := func(a, b int) int {
        if a < b {
            return -1
        }
        if a > b {
            return 1
        }
        return 0
    }

    resultFunc := func(delta diff.Delta, item int) error {
        symbol := map[diff.Delta]string{diff.OLD: "<", diff.NEW: ">"}[delta]
        fmt.Printf("%s %d\n", symbol, item)
        return nil
    }

    result, err := diff.Generic(
        context.Background(),
        streamA, streamB,
        errA, errB,
        compareFunc,
        resultFunc,
    )

    if err != nil {
        panic(err)
    }

    fmt.Printf("Differences found: %d\n", result.ExtraA+result.ExtraB)
}
Parallel Diff Processing

import (
    "context"
    "fmt"
    "sync"

    "github.com/lanrat/extsort/diff"
)

func main() {
    streamA := make(chan string, 100)
    streamB := make(chan string, 100)
    errA := make(chan error, 1)
    errB := make(chan error, 1)

    // Populate streams with test data
    go func() {
        defer close(streamA)
        defer close(errA)
        for i := 0; i < 50; i += 2 {
            streamA <- fmt.Sprintf("item_%03d", i)
        }
    }()

    go func() {
        defer close(streamB)
        defer close(errB)
        for i := 1; i < 50; i += 2 {
            streamB <- fmt.Sprintf("item_%03d", i)
        }
    }()

    // Use channel-based result processing for parallel handling
    resultFunc, resultChan := diff.StringResultChan()

    var wg sync.WaitGroup
    wg.Add(1)

    // Process results in parallel
    go func() {
        defer wg.Done()
        for result := range resultChan {
            fmt.Printf("Difference: %s %s\n", result.D, result.S)
        }
    }()

    // Start diff operation
    go func() {
        defer close(resultChan)
        _, err := diff.Strings(
            context.Background(),
            streamA, streamB,
            errA, errB,
            resultFunc,
        )
        if err != nil {
            fmt.Printf("Diff error: %v\n", err)
        }
    }()

    wg.Wait()
    fmt.Println("Diff processing complete")
}
Performance Considerations

Memory Usage: Configure ChunkSize based on available memory (larger chunks = less I/O, more memory)
Parallelism: Increase NumWorkers on multi-core systems
Temporary Storage:
Explicitly set TempFilesDir to a known disk-backed directory for large datasets
On Linux, prefer /var/tmp over /tmp (which may be tmpfs/memory-backed)
Use fast storage (SSD recommended) for temporary files
Channel Buffers: Tune buffer sizes based on your producer/consumer patterns
Error Handling

The library uses Go's standard error handling patterns. Always check the error channel:

sorter, outputChan, errChan := extsort.Ordered(inputChan, nil)
go sorter.Sort(context.Background())

for item := range outputChan {
    // Process sorted item
}

if err := <-errChan; err != nil {
    // Handle error
    log.Fatal(err)
}
Limitations

Not Stable: The sort is not stable (equal elements may change relative order)
Disk Space: Requires temporary disk space approximately equal to input data size
Memory: Minimum memory usage depends on chunk size configuration
License

This project is licensed under the Apache License - see the LICENSE file for details.

Contributing

Contributions are welcome! Please feel free to submit a Pull Request.

Stargate
Go Report Card PkgGoDev Docker

Stargate is both a Go library and a SOCKS5 proxy server that can egress traffic from multiple IP addresses within a subnet. It supports both TCP CONNECT and UDP ASSOCIATE operations, randomly distributing connections across different IP addresses to help avoid rate-limiting and provide load balancing across your available IP range.

This requires the host running stargate to have the subnet routed directly to it.

If you have an IPv6 subnet, stargate can allow you to make full use of it by any program that can speak SOCKS, or you can use the library directly in your Go applications.

Usage

Usage of ./stargate: [OPTION]... CIDR
 CIDR example: "192.0.2.0/24"
OPTIONS:
  -listen string
     listen on specified IP:port (e.g., '127.0.0.1:1337', '127.0.0.1:8080', '[::1]:1080'). (default "127.0.0.1:1080")
  -subnet-size uint
     CIDR prefix length for random subnet proxy (e.g., 64 for /64 IPv6 subnets)
  -test
     run test request on all IPs and exit
  -verbose
     enable verbose logging
  -version
     print version and exit
Stargate operates as a single SOCKS5 proxy server that supports both TCP and UDP protocols, randomly selecting egress IP addresses from your specified CIDR range. This approach is much more memory-efficient and suitable for large IPv6 ranges.

Protocol Support

Stargate supports the full SOCKS5 specification:

TCP CONNECT: Standard TCP proxy connections
UDP ASSOCIATE: UDP packet relay for applications that need UDP support
Both TCP and UDP listeners run simultaneously on the same port, allowing clients to use either protocol as needed.

Test Flag - Preventing IP Address Leakage

IMPORTANT: Before using Stargate in production, always run the test mode first to ensure there are no unintended IP address leaks.

The -test flag performs comprehensive validation by:

Testing HTTP requests from every available IP address in your CIDR range
Verifying that each egress IP matches the intended source address
Detecting binding errors or network misconfigurations
Ensuring no connections leak through unintended IP addresses
Note: When using -subnet-size, the test will validate one randomly selected IP address from each subnet rather than testing every possible IP. For example, with -subnet-size 64 on a /48, it tests one IP per /64 subnet, not every IP in the entire /48. In order to test every possible IP address, do not pass a -subnet-sizeoption when using -test.

Always run this test before production use:

# Test your configuration first - THIS IS CRITICAL!
./stargate -test 192.0.2.0/24

# Only proceed to normal operation after tests pass
./stargate 192.0.2.0/24
The test will fail immediately if any IP address binding issues are detected, preventing potential IP leakage that could compromise your setup.

Subnet Distribution

When used with -subnet-size, the proxy will randomly distribute connections across different subnets within the main CIDR range. For example, with a /48 IPv6 block and -subnet-size 64, connections will be distributed across random /64 subnets, giving you access to multiple /64 networks within your larger allocation.

Examples

Basic Usage

Start a SOCKS5 proxy on the default port (1080) that randomly egresses from IPs in the 192.0.2.0/24 range:

# Always test first!
./stargate -test 192.0.2.0/24

# Run the proxy after tests pass
./stargate 192.0.2.0/24
Custom Listen Address

Start a SOCKS5 proxy listening on a specific IP and port:

./stargate -listen 127.0.0.7:8080 192.0.2.0/24
IPv6 with Large Address Space

Use an IPv6 /64 subnet - this gives you 2^64 possible egress IPs:

./stargate -test 2001:DB8:1337::/64
./stargate 2001:DB8:1337::/64
Subnet-Level Distribution

Distribute connections across multiple /64 subnets within a /48 IPv6 allocation:

./stargate -test -subnet-size 64 2001:DB8:1337::/48
./stargate -subnet-size 64 2001:DB8:1337::/48
This will randomly select from different /64 networks within your /48, providing both IP and subnet-level distribution.

Routing Setup

Before Stargate can use IP addresses from your subnet, you need to configure your network to route that subnet to the host running Stargate.

Step 1: Configure subnet routing First, ensure your network infrastructure routes the entire subnet to your host machine. This is typically done at your router or network provider level.

Step 2: Add local route Once the subnet is routed to your host, tell your operating system that it can bind to any IP in that range by adding a local route:

# Example: if 192.0.2.0/24 is routed to interface eth0
ip -4 route add local 192.0.2.0/24 dev eth0
Important: Don't assign individual IPs from the subnet directly to your network interface. This prevents the kernel from reserving addresses as broadcast addresses and keeps them available for Stargate.

Step 3: Enable non-local binding (if needed) Some systems require enabling non-local binding to allow applications to bind to IPs that aren't directly assigned to interfaces:

# For IPv4
sysctl net.ipv4.ip_nonlocal_bind=1
# For IPv6
sysctl net.ipv6.ip_nonlocal_bind=1
Download

Precompiled Binaries

The easiest way to get started is with precompiled binaries available for multiple platforms including Linux and FreeBSD. These are statically linked and ready to run without additional dependencies.

Docker

Docker is particularly useful for deployment in containerized environments, though network configuration requires special attention to ensure proper subnet routing.

Running in docker will require --net=host, or the subnet must be routed directly to the container.

docker pull ghcr.io/lanrat/stargate:latest
Building

Building from source is straightforward - just run make!

git clone https://github.com/lanrat/stargate.git
cd stargate
make
This will produce a statically linked binary that's ready to use.

Platforms

Stargate makes use of freebind, which allowes for creating a socket with a source IP address on an interface that does not have that IP directly bound to it. This is only available on Linux and FreeBSD. Stargate can work on other platforms, but it will require every address to be previously bound to the interface before running.

Go Library

You can use Stargate as a Go library in your applications to make HTTP requests or other network connections using random source IP addresses:

package main

import (
    "context"
    "fmt"
    "net/http"
    "net/netip"
    "time"

    "github.com/lanrat/stargate"
)

func main() {
    // Parse your CIDR range
    prefix, err := netip.ParsePrefix("192.0.2.0/24")
    if err != nil {
        panic(err)
    }

    // Create a RandomIPDialer for the subnet
    dialer, err := stargate.NewRandomIPIterator(prefix, 32)
    if err != nil {
        panic(err)
    }

    // Create an HTTP client that uses random source IPs
    client := &http.Client{
        Transport: &http.Transport{
            DialContext: dialer.Dial,
        },
        Timeout: 10 * time.Second,
    }

    // Make requests - each will use a different random IP
    for i := 0; i < 5; i++ {
        resp, err := client.Get("https://httpbin.org/ip")
        if err != nil {
            fmt.Printf("Request %d failed: %v\n", i+1, err)
            continue
        }
        resp.Body.Close()
        fmt.Printf("Request %d: Status %d\n", i+1, resp.StatusCode)
    }
}
You can also get the next random IP and DialFunc separately:

// Get next random IP and dialer function
ip, dialFunc, err := dialer.NextDial()
if err != nil {
    panic(err)
}

fmt.Printf("Next egress IP will be: %s\n", ip)

// Use the dial function directly
conn, err := dialFunc(context.Background(), "tcp", "example.com:80")
if err != nil {
    panic(err)
}
defer conn.Close()
